### PR TITLE
document old way of checking validity of CSRF token

### PR DESCRIPTION
**Pull request by snoek09**
| Q | A |
| --- | --- |
| Doc fix? | yes |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | Related to #4668 |

### CODE REVIEW

**Comment by xabbuh** [comment]
:+1:

File: book/controller.rst [hunk 1]
```diff
@@ -794,6 +794,22 @@ Just like when creating a controller for a route, the order of the arguments of
 order of the arguments, Symfony will still pass the correct value to each
 variable.
 
+Checking the Validity of a CSRF Token
+-------------------------------------
+
+Sometimes you want to use CSRF protection in an action where you don't want to use a
+Symfony form.
+
+If, for example, you're doing a DELETE action, you can use the
```
Comment (OskarStark) [inline book/controller.rst]: DELETE in backticks?

File: book/controller.rst [hunk 2]
```diff
@@ -794,6 +794,22 @@ Just like when creating a controller for a route, the order of the arguments of
...
+:method:`Symfony\\Component\\Form\\Extension\\Csrf\\CsrfProvider\\CsrfProviderAdapter::isTokenValid`
+method to check the CSRF token::
+
+    use Symfony\Component\Security\Csrf\CsrfToken;
+
+    $this->get('security.csrf.token_manager')->isTokenValid(
+        new CsrfToken('token_id', 'TOKEN')
```
Comment (OskarStark) [inline book/controller.rst]: Would use a real secret instead of the string TOKEN

File: book/controller.rst (hunk 1 above)
Comment (snoek09) [inline book/controller.rst]: I checked the docs for mentions of HTTP methods. In a sentence they are never in backticks.

Comment (OskarStark) [inline book/controller.rst]: :+1:

File: book/controller.rst [hunk 3]
```diff
@@ -794,6 +794,22 @@ Just like when creating a controller for a route, the order of the arguments of
...
+Sometimes you want to use CSRF protection in an action where you don't want to use a
+Symfony form.
+
+If, for example, you're doing a DELETE action, you can use the
+:method:`Symfony\\Component\\Form\\Extension\\Csrf\\CsrfProvider\\CsrfProviderAdapter::isTokenValid`
+method to check the CSRF token::
+
+    use Symfony\Component\Security\Csrf\CsrfToken;
+
+    $this->get('security.csrf.token_manager')->isTokenValid(
```
Comment (stof) [inline book/controller.rst]: this service only exists in Symfony 2.4+

File: book/controller.rst [hunk 4]
```diff
@@ -794,6 +794,22 @@ Just like when creating a controller for a route, the order of the arguments of
...
+-------------------------------------
+
+Sometimes you want to use CSRF protection in an action where you don't want to use a
+Symfony form.
+
+If, for example, you're doing a DELETE action, you can use the
+:method:`Symfony\\Component\\Form\\Extension\\Csrf\\CsrfProvider\\CsrfProviderAdapter::isTokenValid`
+method to check the CSRF token::
+
+    use Symfony\Component\Security\Csrf\CsrfToken;
```
Comment (stof) [inline book/controller.rst]: The security-csrf component does not exist in 2.3

File: book/controller.rst (hunk 3 above)
Comment (snoek09) [inline book/controller.rst]: Thanks, will change this to `form.csrf_provider`.

File: book/controller.rst (hunk 4 above)
Comment (snoek09) [inline book/controller.rst]: Thanks, will change this to use `generateCsrfToken` from `Symfony\Component\Form\Extension\Csrf\CsrfProvider`.

**Comment by snoek09** [comment]
@xabbuh please let me know if the example code is fine now.

**Comment by snoek09** [comment]
@xabbuh is this PR finished?

File: book/controller.rst [hunk 5]
```diff
@@ -794,6 +794,22 @@ Just like when creating a controller for a route, the order of the arguments of
...
 variable.
 
+Checking the Validity of a CSRF Token
+-------------------------------------
+
+Sometimes you want to use CSRF protection in an action where you don't want to use a
+Symfony form.
+
+If, for example, you're doing a DELETE action, you can use the
+:method:`Symfony\\Component\\Form\\Extension\\Csrf\\CsrfProvider\\DefaultCsrfProvider::isCsrfTokenValid`
```
Comment (xabbuh) [inline book/controller.rst]: I think this should link to the method in the interface instead (it doesn't matter which implementation is registered as the service).

**Comment by xabbuh** [comment]
@snoek09 Sorry for missing your last comment. I left a last minor comment. After that this looks good to be merged to me.

**Comment by snoek09** [comment]
@xabbuh No problem. I updated the link.

**Comment by OskarStark** [comment]
:+1:

**Comment by snoek09** [comment]
@xabbuh I think this is ready to be merged.

**Comment by xabbuh** [comment]
@weaverryan @WouterJ @javiereguiluz Ready to go here? :)

File: book/controller.rst [hunk 6]
```diff
@@ -794,6 +794,22 @@ Just like when creating a controller for a route, the order of the arguments of
...
+
+Sometimes you want to use CSRF protection in an action where you don't want to use a
+Symfony form.
+
+If, for example, you're doing a DELETE action, you can use the
+:method:`Symfony\\Component\\Form\\Extension\\Csrf\\CsrfProvider\\CsrfProviderInterface::isCsrfTokenValid`
+method to check the CSRF token::
+
+    $csrf = $this->container->get('form.csrf_provider');
+    $intention = 'authenticate';
```
Comment (weaverryan) [inline book/controller.rst]: Does it make sense to show this without showing a template? When you generate the CSRF token in the template, you would need to also include `authenticate` then, right?

File: book/controller.rst [hunk 7]
```diff
@@ -794,6 +794,22 @@ Just like when creating a controller for a route, the order of the arguments of
...
+    $token = $csrf->generateCsrfToken($intention);
+
+    $csrf->isCsrfTokenValid($intention, $token);
```
Comment (weaverryan) [inline book/controller.rst]: How about an `if` statement to be more realistic:

```php
if (!$csrf->isCsrfTokenValid($intention, $token)) {
    // CSRF token invalid! Do something, like redirect with an error
}
```

**Comment by weaverryan** [comment]
I added some comments, but I like it! Short and simple.

To be clear, this is not an "old way" of checking CSRF, correct? It seems totally valid and current.

Thanks!

**Comment by ogizanagi** [comment]
@weaverryan : No it isn't, because the `form.csrf_provider` service is deprecated since 2.4, as well as the whole [`CsrfProviderInterface`](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Form/Extension/Csrf/CsrfProvider/CsrfProviderInterface.php#L30-L32). We use `security.csrf.token_manager` instead.

However, I wonder if the `book/controller.rst` page is the right place for this ? In #4668, @WouterJ suggested to create a new entry in the cookbook instead, and to me it actually makes much more sense to have this in a Controller or Security cookbook section.
I know PRs for 2.6+ are already merged (#5325), but what do you think ?

**Comment by snoek09** [comment]
@ogizanagi is right. The example code has to work with 2.3 using the 'old way'.
I also agree that this documentation should be moved to the cookbook.

**Comment by xabbuh** [comment]
I think we can merge it just as it is and then work on converting this into a cookbook recipe afterwards. What do you think?

**Comment by ogizanagi** [comment]
:+1:

**Comment by snoek09** [comment]
I like that idea @xabbuh.

**Comment by xabbuh** [comment]
@weaverryan @WouterJ @javiereguiluz Okay with merging and thinking about how to rework the whole part afterwards?

**Comment by javiereguiluz** [comment]
@xabbuh yes!

**Comment by weaverryan** [comment]
:+1:

**Comment by xabbuh** [comment]
Thank you Henry.

**Pull request closed without merging**